[@types/underscore] Fix #36308 - underscore has bad chain typings fix only for function 'map' (#36319)

Radullus opened the pull request:
Fix typing in 'underscore' for function 'map'

Please fill in this template.

- [ ] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [ ] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: https://underscorejs.org/#map
https://underscorejs.org/#chain
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

Radullus commented:
@jgonggrijp @regevbr @renjfk  Can you please review?

jgonggrijp commented:
I defer this one to @regevbr because he is currently working on a complete fix for the chain typings. I'm not sure whether the fix in the current PR will combine well with his work, but I think the unittest would be a welcome addition in any case.

regevbr commented on types/underscore/index.d.ts:
```diff
@@ -5094,25 +5094,25 @@ declare module _ {
         * @see _.map
         **/
-        map<TArray>(iterator: _.ListIterator<T, TArray[]>, context?: any): _ChainOfArrays<TArray>;
+        map<TArray>(iterator: _.ListIterator<T, TArray>, context?: any): _ChainOfArrays<TArray>;
```
this should be reverted as current behavior is ok

Radullus replied:
done.
I have add test for this definition of 'map', but I have to comment part of the 'testing' witch check output because it will cause the test to fail.

regevbr commented on types/underscore/index.d.ts:
```diff
@@ -5094,25 +5094,25 @@ declare module _ {
         * @see _.map
         **/
-        map<TArray>(iterator: _.ObjectIterator<T, TArray[]>, context?: any): _ChainOfArrays<TArray>;
+        map<TArray>(iterator: _.ObjectIterator<T, TArray>, context?: any): _ChainOfArrays<TArray>;
```
this should be reverted as current behavior is ok

Radullus replied:
done.
I have add test for this definition of 'map', but I have to comment part of the 'testing' witch check output because it will cause the test to fail.

Radullus commented:
- I added more test, for mapping 'Dictionary'.
 - I apply charges suggested for Line 5103 in line DefinitelyTyped/types/underscore/index.d.ts d200340
 - I modified Line 5115 DefinitelyTyped/types/underscore/index.d.ts d200340. 
Map function for 'chain' object don't return 'Dictionary' but return 'Array'.  You can see some of my tested https://jsfiddle.net/Radullus/fsmj2rxa/37/

regevbr commented on types/underscore/underscore-tests.ts:
```diff
@@ -209,6 +209,28 @@ _(stooges)
+    'fake id': { name: 'curly', age: 60 },
+};
+
+let youngPeopleId: string[] = _.chain(usersData)
```
This test should actually fail, as performing filter on a a Dictionary<T> returns T[], thus performing map on it will need the following iterator:
```
(value: T, index: number, list: T[]): TResult
```
The reason this test passes now is because of missing signature of filter for _Chain which should be:
```ts
filter(iterator: _.ObjectIterator<T, boolean>, context?: any): _Chain<T, T[]>;
```
But fixing it will not even help as the entire _Chain interface is malformed :-( and I'm working on a wide fix for it.
I would rather you only add tests that use the fixed map method

Radullus replied:
You have right. I will stick in this test to 'map' function. I interface definition I was trying to test.
And, I will add some test for 'filter' separate.

regevbr commented:
Thanks @Radullus!
The _Chain interface is completely messed up currently :(
I added some comments for you to review

Radullus commented:
Hi @regevbr . I have applied the changes you had suggested.

regevbr commented:
I added new comment for you to review

regevbr commented on types/underscore/underscore-tests.ts:
```diff
@@ -209,6 +209,46 @@ _(stooges)
+    .value();
+
+// Test map function with _ChainOfArrays<>
+let usersTable_2 /*: { age: number; name: string; id: string }[][]*/ = _.chain(usersData)
```
why did you remove the type here?
It is best to use the flatten() method here and then get the value

Radullus replied:
because the type by the underscore should be like this one in the comment, But with current typescript definition, There is missing one level of array nest. 
`{ age: number; name: string; id: string }[][]  // type return by underscore
{ age: number; name: string; id: string }[]    // type with is calculate base of current definitions
`
It would require more changes in the underscore typescript definition. And I'm not certain what should be changed?
`interface _ChainOfArrays<T> extends _Chain<T[]> `
to 
`interface _ChainOfArrays<T> extends _Chain<T[], T[][]> `
it will not match with data return from function 'groupBy'

regevbr replied:
You are correct. Lets leave it be for now

regevbr commented on types/underscore/underscore-tests.ts:
```diff
@@ -209,6 +209,46 @@ _(stooges)
+    })
+    .value();
+
+let usersTable_3 /*: { score: number; fullName: string; login: string }[][]*/ = _.chain(usersTable)
```
same here - why did you remove the type?
